src/Nethermind/Nethermind.Synchronization/SyncServer.cs

```diff
@@ -249,34 +254,70 @@ private void SyncBlock(Block block, ISyncPeer? syncPeer)
             {
                 if (!_blockValidator.ValidateSuggestedBlock(block))
                 {
-                    string message = $"Peer {syncPeer?.Node:c} sent an invalid block.";
+                    string message = $"Peer {nodeWhoSentTheBlock?.Node:c} sent an invalid block.";
                     if (_logger.IsDebug) _logger.Debug(message);
-                    lock (_recentlySuggested)
-                    {
-                        _recentlySuggested.Delete(block.Hash!);
-                    }
-
+                    _recentlySuggested.Delete(block.Hash!);
                     throw new EthSyncException(message);
                 }
 
-
-
                 bool shouldSkipProcessing = _blockTree.Head.IsPoS() || block.IsPostMerge;
                 if (shouldSkipProcessing)
                 {
                     if (_logger.IsInfo) _logger.Info($"Skipped processing of discovered block {block}, block.IsPostMerge: {block.IsPostMerge}, current head: {_blockTree.Head}");
                 }
 
-                if (_logger.IsTrace) _logger.Trace($"SyncServer SyncPeer {syncPeer} SuggestBlock BestSuggestedBlock {_blockTree.BestSuggestedBody}, BestSuggestedBlock TD {_blockTree.BestSuggestedBody?.TotalDifficulty}, Block TD {block.TotalDifficulty}, Head: {_blockTree.Head}, Head: {_blockTree.Head?.TotalDifficulty}  Block {block.ToString(Block.Format.FullHashAndNumber)}");
+                if (_logger.IsTrace) _logger.Trace($"SyncServer SyncPeer {nodeWhoSentTheBlock} SuggestBlock BestSuggestedBlock {_blockTree.BestSuggestedBody}, BestSuggestedBlock TD {_blockTree.BestSuggestedBody?.TotalDifficulty}, Block TD {block.TotalDifficulty}, Head: {_blockTree.Head}, Head: {_blockTree.Head?.TotalDifficulty}  Block {block.ToString(Block.Format.FullHashAndNumber)}");
                 AddBlockResult result = _blockTree.SuggestBlock(block, shouldSkipProcessing ? BlockTreeSuggestOptions.None : BlockTreeSuggestOptions.ShouldProcess);
                 if (_logger.IsTrace) _logger.Trace($"SyncServer block {block.ToString(Block.Format.FullHashAndNumber)}, SuggestBlock result: {result}.");
             }
             else
             {
-                if (_logger.IsDebug) _logger.Debug($"Peer {syncPeer} sent block with unknown parent {block}, best suggested {_blockTree.BestSuggestedHeader}.");
+                if (_logger.IsDebug) _logger.Debug($"Peer {nodeWhoSentTheBlock} sent block with unknown parent {block}, best suggested {_blockTree.BestSuggestedHeader}.");
             }
         }
 
+        private void BroadcastBlock(Block block, bool allowHashes, ISyncPeer? nodeWhoSentTheBlock = null)
```

Comment:
so as I understand we will broadcast the block two times.
1. After NewBlockMessage
2. After we marked it as a head

Reply:
Yes, this is a compromise.
1. After NewBlockMessage we will broadcast full block to sqrt(peers)
2. After becoming head we will broadcast full block to sqrt(peers) and hash to the rest.

In theory we potentially shouldn't broadcast full blocks again after head, but we would need somehow to store this information. Geth stores in a block information about the peer it came from. We could do that too.

But I didn't found any major drawbacks of broadcasting it again to this fairly small set of peers, so... I ignored it for now.